[deckhouse-controller] feat: add LTS release channel

juev opened the pull request:
## Description

Fixed: #10980

## Why do we need it, and what problem does it solve?

We need to add a new release channel that updates only for very stable versions.

## What is the expected result?

A new release channel has been added, and when a new version is published to this channel, we can update in accordance with the selected update mode.

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: deckhouse-controller
type: feature
summary: add LTS release channel
impact_level: default
```

juev commented:
How I tested:

<img width="704" alt="Screenshot 2024-12-09 at 16 57 45" src="https://github.com/user-attachments/assets/e3de2499-0e2c-4552-af69-e73548c8c27e">

Brought the mc deckhouse to the view:

```yaml
apiVersion: deckhouse.io/v1alpha1
kind: ModuleConfig
metadata:
  finalizers:
  - modules.deckhouse.io/module-config
  name: deckhouse
spec:
  enabled: true
  settings:
    bundle: Default
    logLevel: Info
    releaseChannel: LTS
    update:
      mode: Auto
  version: 1
```

New value for the field `releaseChannel: LTS`.

After that, I created the 1.77 release:
```yaml
apiVersion: deckhouse.io/v1alpha1
approved: false
kind: DeckhouseRelease
metadata:
  annotations:
    dryrun: "true"
    release.deckhouse.io/approved: "true"
  name: v1.77.0
spec:
  version: v1.77.0
```

И затем создал релиз 1.88:
```yaml
apiVersion: deckhouse.io/v1alpha1
approved: false
kind: DeckhouseRelease
metadata:
  annotations:
    dryrun: "true"
    release.deckhouse.io/approved: "true"
  name: v1.88.0
spec:
  version: v1.88.0
```

As you can see from the console output, the releases have been successfully installed.

deckhouse-BOaTswain commented:
Aye, aye, @juev. I've started the workflow for label 'e2e/run/aws'...
<!-- workflow_start -->

juev commented:
The current version is tested as follows:
1. Manually created release v1.01.0
2. Pushed version v1.69.1 into the dev registry
3. In the logs, I waited until the check for new releases started and saw the creation of a new release
4. Result

<img width="527" alt="Screenshot 2024-12-23 at 11 25 52" src="https://github.com/user-attachments/assets/1109c6dd-9050-411e-9de7-98a65d829cc6" />

juev commented:
How did I test?

I uploaded a number of versions to the dev registry and created Stable and LTS release channels for testing:

<img width="652" alt="Screenshot 2024-12-24 at 15 19 35" src="https://github.com/user-attachments/assets/6241a37e-b518-4fd6-a2ed-b1220707c0bf" />

I then switched DH to Stable release channel and Manual update mode and waited for the next registry check to pass. Made sure the required DH versions were created:

<img width="927" alt="Screenshot 2024-12-24 at 15 16 55" src="https://github.com/user-attachments/assets/13fd29d8-5486-442f-a7f2-96b9106c30b6" />

Then I deleted the created versions, switched DH to the LTS release channel, and waited for the registry to check. I made sure that only the version specified in the release channel is being created:

<img width="930" alt="Screenshot 2024-12-24 at 15 20 08" src="https://github.com/user-attachments/assets/9daaaffd-abcd-47d5-a481-19a2116df9f4" />

ldmonster commented on deckhouse-controller/pkg/controller/deckhouse-release/check_release.go:
```diff
@@ -542,51 +541,61 @@ func (dcr *DeckhouseReleaseChecker) StepByStepUpdate(ctx context.Context, actual
-				continue
+	var patchVersion, nextVersion *semver.Version
+	for _, tag := range vs {
+		if tag.Major() == actual.Major() &&
```
we can check for major or minor. if no - it's a patch version.
do not need in double check.

juev replied:
We cannot interrupt the check for nextVersion, since we take the last available patch there. As a result, if we do not check major/minor here, a collision may occur, that is, we will take the patch version from a higher version.

juev replied:
done

ldmonster commented on deckhouse-controller/pkg/controller/deckhouse-release/check_release.go:
```diff
@@ -542,51 +541,61 @@ func (dcr *DeckhouseReleaseChecker) StepByStepUpdate(ctx context.Context, actual
-		return nil, err
+	var vs []*semver.Version
+	for _, tag := range tags {
+		if tag.Compare(actual) >= 0 && tag.Compare(target) <= 0 {
```
explain purpose of adding tag equal to actual please.

juev replied:
fixed

ldmonster commented on deckhouse-controller/pkg/controller/deckhouse-release/check_release.go:
```diff
@@ -542,51 +541,61 @@ func (dcr *DeckhouseReleaseChecker) StepByStepUpdate(ctx context.Context, actual
+				(tag.Major() == nextVersion.Major() &&
+					tag.Minor() == nextVersion.Minor() &&
+					tag.Patch() > nextVersion.Patch()) {
+				nextVersion = tag
```
break here to return only first tag with minor/major?

juev replied:
We return not the first tag we come across, but the last patch for major/minor, as a result, until we go to the end, we cannot understand whether we have checked all the patches or not.

ldmonster commented on deckhouse-controller/pkg/controller/deckhouse-release/check_release.go:
```diff
@@ -611,13 +620,22 @@ func (dcr *DeckhouseReleaseChecker) generateChangelogForEnabledModules() map[str
+		for _, tag := range tags {
+			version, err := semver.NewVersion(tag)
+			if err != nil {
+				continue
```
log error?

ldmonster replied:
```suggestion
				dcr.log.Warn("bad semver", slog.String("version", tag))
```

ldmonster commented on deckhouse-controller/pkg/controller/deckhouse-release/check_release.go:
```diff
@@ -611,13 +620,22 @@ func (dcr *DeckhouseReleaseChecker) generateChangelogForEnabledModules() map[str
-		dcr.tags, err = dcr.registryClient.ListTags(ctx)
+		tags, err := dcr.registryClient.ListTags(ctx)
+		if err != nil {
+			return nil, err
```
```suggestion
			return nil, fmt.Errorf("registy client list tags: %w",err)
```
